src/Expr.cc

```diff
@@ -4644,62 +4644,77 @@ void CallExpr::ExprDescribe(ODesc* d) const
 		args->Describe(d);
 	}
 
-LambdaExpr::LambdaExpr(std::unique_ptr<FunctionIngredients> arg_ing, IDPList arg_outer_ids,
-                       StmtPtr when_parent)
+LambdaExpr::LambdaExpr(std::shared_ptr<FunctionIngredients> arg_ing, IDPList arg_outer_ids,
```

Comment:
Copying from our brief Slack discussion:

For that one, I think there should be broader discussion, as to me it seems undesirable to derive from a class just to use Zeek’s quirky home-grown smart pointer, when standard ones work just as well, and are (or at least eventually should be) more idiomatic.

Reply:
(That said, the thread safety guarantees weren't on my radar. I take it there isn't a more streamlined version of `shared_ptr` that doesn't provide those?)

Reply:
> as to me it seems undesirable to derive from a class just to use Zeek’s quirky home-grown smart pointer, when standard ones work just as well and are (or at least eventually should be) more idiomatic.

I saw `Frame` and some others derived from `zeek::Obj`, so seemed fair to suggest. To me IntrusivePtr is mostly a shared_ptr with references managed within the object (probably oversimplifying), so not finding it so quirky and its use is pervasive.

Mostly this was coming from the consistency side of types of pointer received for Expr/Stmt. We currently have a mix of raw pointers and IntrusivePtr. Here and there a unique_ptr or shared_ptr is popping up now.

Reply:
Okay, similar to the naming-of-arguments-to-constructors issue discussed above with @timwoj, once there's a definitive style ruling I'll go ahead and accommodate it.

FWIW, `Frame` is derived from `Obj` because it's very old code, dating to when there was no standard support for reference-counting pointers. Personally I like the notion of migrating Zeek over to standard smart pointers, as with time it'll help with newcomers familiarizing themselves with the code base. But I appreciate this is a matter of taste, and the question of consistent style is a legitimate one.

Reply:
> Personally I like the notion of migrating Zeek over to standard smart pointers, as with time it'll help with newcomers familiarizing themselves with the code base. But I appreciate this is a matter of taste, and the question of consistent style is a legitimate one.

Yeah, I swinging towards that direction and start to be neutral. Had chatted a bit with Robin..

If we were to place one or two `using FunctionIngredientsPtr = shared_ptr<FunctionIngredients>` into the headers and use that rather the long and explicit type, that actually alleviates my "looking odd" concern in fact. (AnalyzerPtr, EventGroupPtr being examples of that).

Could have `Func::AddBody()` then using the `const FunctionIngredientsPtr&` and could skip the de-reference part at the caller.

Could you see that? If not, okay with merging as is and discuss outside of the PR or in a follow-up.

Reply:
I like `using FunctionIngredientsPtr = shared_ptr<FunctionIngredients>`, that then informs the developer "it's a smart pointer, no need to worry about the particulars", and fits with the existing style. I'll add that to the upcoming push.

Reply:
Ah - it was changed in the header, but the final version still uses std::shared_ptr here.

Will fix during merge.